File: src/tusk/interpreter.py
import re
import math
import asyncio
import json
import os

from logger import *
from tusk.lexer import *

from tusk.nodes.expressions import FactorNode, TermNode, ExpressionNode
from tusk.nodes.statement import StatementNode
from tusk.nodes.base.function import FunctionNode
from tusk.nodes.base.return_node import ReturnNode

### INTERPRETER

class Interpreter:
    def __init__(self):      
        self.return_value = True

    def setup(self, data=None, tokens=None, text=None, file=None, ext=[]):
        ## Lang Init
        with open("lang/main.json","r") as f:
            main = json.load(f)
        for i in ext:
            # load the extension
            with open(i,"r") as f:
                e = json.load(f)
            for name, conf in e["lang"]["lexer"].items():
                main["lang"]["lexer"][name] = conf
        

        ## Interpreter Init
        if data==None:
            self.data = {
                "vars":{},
                "funcs":{},
                "local":{},
                "async_tasks":[],
            }
        else:
            self.data = data
        if text!=None:self.text=text
        if file!=None:
            with open(file, "r") as f:
                self.text = f.read()
            self.file=file
        else: self.file = "<stdin>"

        if tokens==None:
            self.tokens = Lexer(self.text, self).classify_tokens()
        else:
            self.tokens = tokens 
            self.tokens = self.change_token_parent(self)                       
            

        self.pos = 0
        self.current_token = self.tokens[self.pos]

        with open('config.json', 'r') as f:
            self.config = json.load(f)
        self.debug = self.config["debug"]
        self.debug_msg(self.tokens)
        return self

    async def compile(self):
        self.end_found = False
        self.caught_error = False
        while self.pos <= len(self.tokens)-1:
            #self.debug_msg(self.current_token, "<- stmt start")
            if self.end_found:
                self.debug_msg("RETURN ENDSCRIPT")
                return self.return_value
            if self.current_token.type == "ENDSCRIPT":
                self.debug_msg("DEFAULT ENDSCRIPT", "<- stmt end")
                return self.return_value
            elif self.current_token.type == "NEWLINE":
                self.next_token()
                continue
            elif self.current_token.type == "BREAKSTRUCTURE" and self.current_token.value == "return":
                await ReturnNode(self.current_token).create()
                break
            else:
                
                if self.current_token.type == "ENDSCRIPT": 
                    self.debug_msg("FAILCHECK ENDSCRIPT", "<- stmt end")
                    return self.return_value
                try:
                    await StatementNode(self.current_token).create()
                except Exception as e:
                    self.error("UnknownError", str(e))
                    raise e
            self.debug_msg(self.current_token, "<- stmt end")
            if self.get_next_token() == None: 
                self.debug_msg("MISS ENDSCRIPT", "<- stmt end")
                break
            else: 
                e = self.next_token()
                if e.type == "ENDSCRIPT": 
                    self.debug_msg("DEFAULT ENDSCRIPT", "<- stmt end")
                    return self.return_value


    def change_token_parent(self, interpreter) -> list[Token]:
        tokens = []
        for token in self.tokens:
            token.interpreter = interpreter
            tokens.append(token)
        return tokens
    
    def get_var(self, var_name) -> any:
        if isinstance(var_name, Token): var_name = var_name.value
        if var_name in self.data["vars"]:
            return self.data["vars"][var_name]
        else:
            raise Exception(f"IDENTIFIER {var_name} is undefined")
    

    def arrows_at_pos(self) -> str:
        recreated_code = ""
        arrows=""
        npos = 0
        target=""
        for i in self.tokens:
            npos += 1
            if npos >= self.pos-2 and npos <= self.pos+4:
                # Format token value
                if i.type == "STRING":
                    token_str = f' "{i.value}"'
                    width = len(i.value) + 3
                else:
                    token_str = f" {i.value}"
                    width = len(i.value) + 1
                
                recreated_code += token_str
                
                # Add arrows
                if npos == self.pos+1:
                    arrows += "^" * width
                    target = i
                else:
                    arrows += " " * width
        recreated_code+=f"<---- {str(target)}"
                
        return f"{recreated_code}\n{arrows}"
            


    def get_next_token(self,nx=0) -> Token|None: 
        if self.pos+nx >= len(self.tokens)-1:
            return None
        else: return self.tokens[self.pos+1+nx]
    
    def is_token(self,token_type) -> bool:
        token_type = token_type.split(":")
        if self.get_next_token().value == token_type[1] and self.get_next_token().type == token_type[0]:
            return True
        else:
            return False

    def next_token(self) -> Token:
        next_tkn = self.get_next_token()
        if next_tkn !=None:
            self.pos+=1
            self.current_token = self.tokens[self.pos]
            return self.current_token
        else:
            self.error("UnfinishedExpression", "Unfinished expression at ENDSCRIPT")
    
    def expect_token(self, token_types, notes:list[str]=[]) -> Token:
        token_types = token_types.replace(" ","").split("|")

        next_tkn = self.get_next_token()
        self.debug_msg(next_tkn, "<- to be expected token")
        for i in token_types:
            if ":" in i:
                i = i.split(":")[1]
                if next_tkn.value == i:
                    return self.next_token()
            else:
                if next_tkn.type == i:
                    return self.next_token()
        if "IDENTIFIER" in token_types:
            self.error("UnexpectedToken",f"Expected token {str(token_types)}, got {next_tkn}", notes=["Possible Fix: You might have entered a keyword as a variable name, try renaming it"])
        else:
            self.error("UnexpectedToken",f"Expected token {str(token_types)}, got {next_tkn}",notes=notes)
    
    def expect_tokens(self, token_types) -> list[Token]:
        token_types = token_types.replace(" ","").split(";")
        tokens = []

        for i in token_types:
            if self.get_next_token().type not in i.split("|"):
                raise Exception(f"Expected token {str(token_types)}, got {self.current_token}")
            else: 
                tokens.append(self.get_next_token())
                self.next_token()
        
        return tokens
    
    def error(self, error_name:str, error_desc:str, notes:list[str]=[]):
        if not self.caught_error: # to prevent error spam
            self.caught_error = True
            print("\n================ ERROR ================")
            print(f"{error_name}: {error_desc}")
            print("============== POSITION ===============")
            print(self.arrows_at_pos())
            print("================ NOTES ================")
            for i in notes:
                print(i)
            print("=======================================\n")
            self.end_found = True
            #exit()
        self.caught_error = True

    def debug_msg(self, *args,color="blue"):
        if self.debug: cprint(*args,color=color,engine="debug")
    
    def print_(self,*args,color="normal",engine="tusk"):
        cprint(*args,color=color,engine=f"{engine}@{self.file}")
        



        
            
            

File: src/java/tusan/interpreter/Interpreter.java
package net.tutla.tums.tusan.interpreter;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import net.tutla.tums.tusan.Utils;
import net.tutla.tums.tusan.lexer.Lexer;
import net.tutla.tums.tusan.lexer.Token;
import net.tutla.tums.tusan.lexer.TokenType;
import net.tutla.tums.tusan.nodes.Statement;
import net.tutla.tums.tusan.nodes.base.Return;

public class Interpreter {

    private static Lexer lexer;
    public Utils util = new Utils();

    public InterpreterData data;
    public String text;
    public String file;
    public List<Token> tokens;
    public Object returned;
    public Boolean isFunction = false;

    private Path filePath;

    public Boolean end = false;
    public Boolean caughtError = false;
    public int pos = 0;
    public Token currentToken;


    public Interpreter() {

    }

    public void setup(InterpreterData data, List<Token> _tokens, String _text, Path file){
        this.data = data != null ? data : new InterpreterData(null, null, null, null);

        if (_text != null) {
            this.text = _text;
        }
        if (file != null) {
            this.text = readFileContents(file);
            this.file = String.valueOf(file.toAbsolutePath());
            this.filePath = filePath;
        }
        else {
            this.file = "<stdin>";
        }

        if (_tokens == null){
            lexer = new Lexer(this.text, this);
            this.tokens = lexer.classify();
        } else {
            this.tokens = _tokens;
            this.tokens = changeTokensParent(this);
        }

        this.currentToken = tokens.get(this.pos);
    }

    public Object compile() {
        changeTokensParent(this);
        end = false;
        caughtError = false;
        pos = 0;
        currentToken = tokens.get(pos);

        /* for (Token t : tokens){
            System.out.print(t.type);
            System.out.print(":");
            System.out.print(t.value);
            System.out.print("\n");
        } */
        if (!isFunction){
            System.out.println("================ OUTPUT ===============");
        }
        while (pos <= tokens.toArray().length-1){
            if (end){
                return returned;
            }
            if (currentToken.type == TokenType.ENDSCRIPT){ // how did you get here?
                return returned;
            } else if (currentToken.type == TokenType.BREAKSTRUCTURE && currentToken.value.equals("return")){
                new Return(nextToken()).create();
            } else {
                new Statement(currentToken).create();
            }
            if (getNextToken() == null){
                meetEnd();
            } else {
                Token e = nextToken();
                if (e.type == TokenType.ENDSCRIPT){
                    meetEnd();
                }
            }

        }
        meetEnd();

        return returned;
    }

    // utils

    public Token getNextToken(){
        if (pos >= tokens.toArray().length-1){
            return null;
        } else {
            return tokens.get(pos+1);
        }
    }

    public Token nextToken(){
        Token nxt = getNextToken();
        if (nxt != null){
            this.pos++;
            currentToken = tokens.get(pos);
            return nxt;
        } else {
            error("UnfinishedExpression", "Unfinished expression at ENDSCRIPT", null);
        }
        return null;
    }

    public Token expectTokenType(TokenType token){
        Token nxt = nextToken();
        if (nxt.type == token){
            return nxt;
        } else {
            error("UnexpectedToken", "Expected "+token.name()+" got "+nxt.type.name(), null);
            return null;
        }
    }

    public Token expectToken(TokenType token, String name){
        Token nxt = nextToken();
        if (nxt.type == token && nxt.value.equals(name)){
            return nxt;
        } else {
            error("UnexpectedToken", "Expected "+token.name()+":"+name+" got "+nxt.type.name()+":"+nxt.value, null);
            return null;
        }
    }

    public Token expectTokenClassic(String tokenTypes) {
        String[] types = tokenTypes.replace(" ", "").split("\\|");
        Token nextTkn = getNextToken();

        for (String t : types) {
            if (t.contains(":")) {
                String expectedValue = t.split(":")[1];
                if (expectedValue.equals(nextTkn.value)) {
                    return nextToken();
                }
            } else {
                if (nextTkn.type.toString().equalsIgnoreCase(t)) {
                    return nextToken();
                }
            }
        }

        if (Arrays.asList(types).contains("IDENTIFIER")) {
            error("UnexpectedToken", "Expected token " + Arrays.toString(types) + ", got " + nextTkn,
                    List.of("Possible Fix: You might have entered a keyword as a variable name, try renaming it"));
        } else {
            error("UnexpectedToken", "Expected token " + Arrays.toString(types) + ", got " + nextTkn, null);
        }

        return null;
    }


    public void error(String name, String detail, List<String> notes){
        if (!caughtError){
            System.out.println("================ ERROR ================");
            System.out.println(name+" : "+detail);
            System.out.println("============== POSITION ===============");
            System.out.println(arrowsAtPosition());
            System.out.println("================ NOTES ================");
            if (notes != null){
                for (String note : notes){
                    System.out.println(note);
                }
            }
            System.out.println("=======================================");
            meetEnd();
        }
        caughtError = true;
    }

    public String arrowsAtPosition(){
        StringBuilder recreated = new StringBuilder();
        StringBuilder arrows = new StringBuilder();
        int npos = 0;
        String target = "NOTHING:UNKNOWN";
        for (Token i : tokens){
            npos++;
            if (npos >= pos-2 && npos <= pos+4){
                String tokenStr;
                int width;
                if (i.type == TokenType.STRING){
                    tokenStr = " \"" + i.value + "\"";
                    width = i.value.length() + 3;
                } else {
                    tokenStr = " "+i.value;
                    width = i.value.length()+1;
                }

                recreated.append(tokenStr);
                String tt;
                if (npos == pos+1){
                    tt = "^";
                    target = i.value;
                } else {
                    tt = " ";
                }
                arrows.append(tt.repeat(width));
            }
        }
        recreated.append("\t\t<----- "+target).append("\n").append(arrows);
        return recreated.toString();
    }

    public static String readFileContents(Path path) {
        try {
            return Files.readString(path);
        }
        catch (IOException ex) {
            throw new RuntimeException(ex);
        }
    }

    public List<Token> changeTokensParent(Interpreter interpreter){
        List<Token> s = new ArrayList<>();
        for (Token token : this.tokens){
            token.interpreter = interpreter;
            s.add(token);
        }
        return s;
    }

    public void meetEnd(){
        this.end = true;
        if (!isFunction){
            System.out.println("=======================================");
        }
        // System.exit(0);
    }

    public Interpreter clone(){
        Interpreter intr = new Interpreter();
        intr.setup(data, tokens, text, filePath);
        return intr;
    }
}
